pkg/blockbuilder/blockbuilder.go

```diff
@@ -274,11 +274,72 @@ func (b *BlockBuilder) nextConsumeCycle(ctx context.Context, cycleEnd time.Time)
 		return nil
 	}
 
-	// TODO(v): rebalancing can happen between the calls to consumePartitions; if that happens, the instance may loose
+	// TODO(v): rebalancing can happen between the calls to consumePartition; if that happens, the instance may loose
 	// the ownership of some of its partitions
 	// TODO(v): test for this case
+
+	commitOffsets, err := kadm.NewClient(b.kafkaClient).ListCommittedOffsets(ctx, b.cfg.Kafka.Topic)
+	if err != nil {
+		return fmt.Errorf("get committed offsets: %w", err)
+	}
 	for part, lag := range parts {
-		err := b.consumePartitions(ctx, cycleEnd, part, lag)
+		// We look at the commit offset timestamp to determine how far behind we are lagging
+		// relative to the cycleEnd. We consume the partition in parts accordingly.
+		offset, ok := commitOffsets.Lookup(b.cfg.Kafka.Topic, part)
+		if ok {
+			commitTime := time.UnixMilli(offset.Timestamp)
```

Comment:
On this one

> `kadm.ListedOffset.Timestamp`

I'm not sure the `Timestamp` stored for the committed offset, is the timestamp of the record, this offset pointed at. Can you double-check that in real Kafka?

At least the kgo's [docs say the following][1]

> [..] Timestamp is the millisecond of the offset if listing after a time, otherwise -1.

[1]: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#ListedOffset